internal/metrics/metrics.go

```diff
@@ -0,0 +1,38 @@
+package metrics
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/pkg/errors"
+
+	"github.com/open-policy-agent/opa/internal/metrics/prometheus"
+)
+
+// Metrics is the current global metrics provider
+var Metrics Provider
+
+func init() {
+	Metrics = &dummyProvider{}
+}
+
+// Provider is the API of metric providers which must be implemented by each provider
+type Provider interface {
+	RegisterEndpoints(registrar func(path, method string, handler http.Handler))
+	InstrumentHandler(handler http.Handler, label string) http.Handler
+	Gather() (interface{}, error)
+	Name() string
+}
+
+// SetProvider sets the current global metrics provider singleton by the provider name and its config settings
+func SetProvider(name string, config json.RawMessage) error {
+	switch name {
+	case "":
+		Metrics = &dummyProvider{}
+	case prometheus.ProviderName:
+		Metrics = prometheus.NewPrometheusProvider()
```

Comment:
I'm wondering if a model more like the plugin manager and plugins would work better here. The idea being plugins implementing the Plugin API and in this case the Provider API would register themselves at init() time and then can be looked up by name (in this case the metrics_provider.name config option). In theory we can avoid having to have any imports from the prometheus code leak into this internal metrics one by doing that.

The other benefit is that plugins wanting to implement their own metrics API will get the full config and start/reload stuff that other plugins do.

Reply:
I thought of making this a plugin like other plugins in OPA. However there is an important difference: other plugins can be reconfigured through the API while the server itself depends on the metrics and therefore the metrics provider must be configured in initial config and is not reconfigurable. I felt like having 2 different plugin systems in OPA would be confusing. Especially when we only have just one provider.

Also I see that other built-in plugins are statically imported so not much of a difference here. There is an ability to pass extra settings to the provider. Prometheus provider just doesn't have any.

Regarding registering themselves I felt like it doesn't make much sense where there is only one pseudo-plugin without real dynamically-loaded plugin system. I could change the code to let the provider register itself but will require 3rd package with the singleton registry that would be imported by both `prometheus` and `metrics`. IDK if it worth it right now. Might be better to leave it until we have more providers and use-cases as it can be implemented later without breaking dependencies

Reply:
Also, if it's a real plugin it shouldn't be in `internal` then as we agreed

Reply:
If we go with the approach @tsandall suggested than we can avoid this. I do agree that having it as a full plugin isn't the best option. I think passing in the interface to the server and status plugin helps to solve the main motivations I had for suggesting it.